modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -0,0 +1,549 @@
+                        partsToCollect.remove(pstat.partId());
+                }
+            }
+
+            if (!F.isEmpty(partsToRmv)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Remove local partitioned statistics [key=" + cfg.key() +
+                        ", part=" + partsToRmv + ']');
+                }
+
+                partsToRmv.forEach(p -> {
+                    assert !partsToCollect.contains(p);
+
+                    repo.clearLocalPartitionStatistics(cfg.key(), p);
+                });
+            }
+
+            if (!F.isEmpty(colsToRmv))
+                dropColumnsOnLocalStatistics(cfg, colsToRmv);
+
+            if (!F.isEmpty(partsToCollect))
+                gatherLocalStatistics(cfg, tbl, parts, partsToCollect, colsToCollect);
+            else {
+                // All local statistics by partition are available.
+                // Only refresh aggregated local statistics.
+                gatherer.aggregateStatisticsAsync(cfg.key(), () -> aggregateLocalGathering(cfg.key(), parts));
+            }
+        }
+        catch (Throwable ex) {
+            log.error("Unexpected error on check local statistics", ex);
+        }
+        finally {
+            stopLock.leaveBusy();
+        }
+    }
+
+    /** */
+    private void onChangeStatisticConfiguration(
+        StatisticsObjectConfiguration oldCfg,
+        final StatisticsObjectConfiguration newCfg
+    ) {
+        synchronized (mux) {
+            if (log.isDebugEnabled())
+                log.debug("Statistic configuration changed [old=" + oldCfg + ", new=" + newCfg + ']');
+
+            StatisticsObjectConfiguration.Diff diff = StatisticsObjectConfiguration.diff(oldCfg, newCfg);
+
+            if (!F.isEmpty(diff.dropCols()))
+                dropColumnsOnLocalStatistics(newCfg, diff.dropCols());
+
+            if (!F.isEmpty(diff.updateCols())) {
+                GridH2Table tbl = schemaMgr.dataTable(newCfg.key().schema(), newCfg.key().obj());
+
+                GridCacheContext cctx = tbl.cacheContext();
+
+                Set<Integer> parts = cctx.affinity().primaryPartitions(
+                    cctx.localNodeId(), cctx.affinity().affinityTopologyVersion());
+
+                gatherLocalStatistics(
+                    newCfg,
+                    tbl,
+                    parts,
+                    parts,
+                    diff.updateCols()
+                );
+            }
+        }
+    }
+
+    /** */
+    private void gatherLocalStatistics(
+        StatisticsObjectConfiguration cfg,
+        GridH2Table tbl,
+        Set<Integer> partsToAggregate,
+        Set<Integer> partsToCollect,
+        Map<String, StatisticsColumnConfiguration> colsToCollect
+    ) {
+        if (F.isEmpty(colsToCollect))
+            colsToCollect = cfg.columns();
+
+        Column[] cols = IgniteStatisticsHelper.filterColumns(tbl.getColumns(), colsToCollect.keySet());
+
+        gatherer.gatherLocalObjectsStatisticsAsync(
```

Comment:
If arguments can't fit to a single line, arguments are split by 120 char limit to several lines. But single-line style is preferred...